packages/@aws-cdk/core/lib/private/runtime-info.ts

```diff
@@ -1,95 +1,81 @@
-      if (name.startsWith(`${scope}/`)) {
-        foundMatch = true;
-      }
-    }
-    foundMatch = foundMatch || ALLOWED_PACKAGES.includes(name);
+/**
+ * Source information on a construct (class fqn and version)
+ */
+export interface ConstructInfo {
+  readonly fqn: string;
+  readonly version: string;
+}
 
-    if (!foundMatch) {
-      delete libraries[name];
-    }
+export function constructInfoFromConstruct(construct: IConstruct): ConstructInfo | undefined {
+  const jsiiRuntimeInfo = Object.getPrototypeOf(construct).constructor[JSII_RUNTIME_SYMBOL];
+  if (typeof jsiiRuntimeInfo === 'object'
+    && jsiiRuntimeInfo !== null
+    && typeof jsiiRuntimeInfo.fqn === 'string'
+    && typeof jsiiRuntimeInfo.version === 'string') {
+    return { fqn: jsiiRuntimeInfo.fqn, version: jsiiRuntimeInfo.version };
   }
-
-  // add jsii runtime version
-  libraries['jsii-runtime'] = getJsiiAgentVersion();
-
-  return { libraries };
+  return undefined;
 }
 
 /**
- * Determines which NPM module a given loaded javascript file is from.
- *
- * The only infromation that is available locally is a list of Javascript files,
- * and every source file is associated with a search path to resolve the further
- * ``require`` calls made from there, which includes its own directory on disk,
- * and parent directories - for example:
- *
- * [ '...repo/packages/aws-cdk-resources/lib/cfn/node_modules',
- *   '...repo/packages/aws-cdk-resources/lib/node_modules',
- *   '...repo/packages/aws-cdk-resources/node_modules',
- *   '...repo/packages/node_modules',
- *   // etc...
- * ]
- *
- * We are looking for ``package.json`` that is anywhere in the tree, except it's
- * in the parent directory, not in the ``node_modules`` directory. For this
- * reason, we strip the ``/node_modules`` suffix off each path and use regular
- * module resolution to obtain a reference to ``package.json``.
- *
- * @param fileName a javascript file name.
- * @returns the NPM module infos (aka ``package.json`` contents), or
- *      ``undefined`` if the lookup was unsuccessful.
+ * For a given stack, walks the tree and finds the runtime info for all constructs within the tree.
+ * Returns the unique list of construct info present in the stack,
+ * as long as the construct fully-qualified names match the defined allow list.
  */
-function findNpmPackage(fileName: string): { name: string, version: string, private?: boolean } | undefined {
-  const mod = require.cache[fileName];
-
-  if (!mod?.paths) {
-    // sometimes this can be undefined. for example when querying for .json modules
-    // inside a jest runtime environment.
-    // see https://github.com/aws/aws-cdk/issues/7657
-    // potentially we can remove this if it turns out to be a bug in how jest implemented the 'require' module.
-    return undefined;
+export function constructInfoFromStack(stack: Stack): ConstructInfo[] {
+  function isConstructInfo(value: ConstructInfo | undefined): value is ConstructInfo {
```

Comment:
nit: as a habit I like to use lambdas for inline functions:

```ts
const isConstructInfo = (value: ConstructInfo | undefined): value is ConstructInfo...
```

In method, this has implications on `this`.